crdt/rlist.go

```diff
@@ -926,20 +953,109 @@ func (l *RList) DeleteCharAt(i int) error {
 	return l.DeleteChar()
 }
 
+// +-----------------------------------+
+// | Operations - Insert str container |
+// +-----------------------------------+
+
+//Inserts a string container as a child of the root atom.
+type InsertStr struct{}
+
+func (v InsertStr) AtomPriority() int { return 30 }
+func (v InsertStr) MarshalJSON() ([]byte, error) {
+	return json.Marshal("insert str container")
+}
+
+func (v InsertStr) String() string { return "STR: " }
+
+func (v InsertStr) ValidateChild(child AtomValue) error {
+	switch child.(type) {
+	case InsertChar:
+		return nil
+	case Delete:
+		return nil
+	default:
+		return fmt.Errorf("invalid atom value after InsertStr: %T (%v)", child, child)
+	}
+}
+
+// InsertStr inserts a Str container after the cursor position and advances the cursor.
+func (l *RList) InsertStr() error {
+	l.Timestamp++
+	if l.Timestamp == 0 {
+		// Overflow
+		return ErrStateLimitExceeded
+	}
+	i := siteIndex(l.Sitemap, l.SiteID)
+	atomID := AtomID{
+		Site:      uint16(i),
+		Index:     uint32(len(l.Yarns[i])),
+		Timestamp: l.Timestamp,
+	}
+	atom := Atom{
+		ID:    atomID,
+		Cause: AtomID{}, //Root is the cause atom.
+		Value: InsertStr{},
+	}
+	l.Yarns[i] = append(l.Yarns[i], atom)
+
+	l.insertAtom(atom, 0)
+	l.Cursor = atomID
+	return nil
+}
+
 // +------------+
 // | Conversion |
 // +------------+
 
-// AsString interprets list as a sequence of chars.
-func (l *RList) AsString() string {
+// ToString interprets list as a sequence of chars.
+func (l *RList) ToString() string {
 	atoms := l.filterDeleted()
 	chars := make([]rune, len(atoms))
 	for i, atom := range atoms {
-		chars[i] = atom.Value.(InsertChar).Char
+		if _, ok := atom.Value.(InsertStr); ok {
+			chars[i] = '*'
+		} else if _, ok := atom.Value.(InsertChar); ok {
+			chars[i] = atom.Value.(InsertChar).Char
+		}
 	}
 	return string(chars)
 }
 
+// this interface represents a generic type.
+type generic interface{}
+
+// ToJSON interprets list as a JSON.
+func (l *RList) ToJSON() ([]byte, error) {
+	tab := "    "
+	atoms := l.filterDeleted()
+	genericList := make([]generic, 0, 0)
```

Comment:
You don't need to initalize an empty list, see [this](https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices)

Reply:
BTW, naming in Go is more terse than other languages, so instead of `genericList` prefer something like `elements` or even `xs`, if the function is small enough. Since this is a list intended for JSON, `objects` or `objs` are fine.

Reply:
Changes were made.